src/rust/engine/nailgun/src/lib.rs

```diff
@@ -37,20 +39,119 @@ use std::net::Ipv4Addr;
 use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
 use std::path::PathBuf;
 use std::sync::Arc;
+use tokio::io::AsyncWriteExt;
 
 use bytes::Bytes;
 use futures::channel::{mpsc, oneshot};
-use futures::{future, sink, stream, FutureExt, SinkExt, StreamExt, TryStreamExt};
+use futures::{future, sink, stream, FutureExt, SinkExt, Stream, StreamExt, TryStreamExt};
 use log::{debug, error, info};
 pub use nails::execution::ExitCode;
 use nails::execution::{self, sink_for, stream_for, ChildInput, ChildOutput};
-use nails::{Child, Nail};
+use nails::{Child, Config, Nail};
 use tokio::fs::File;
 use tokio::net::TcpListener;
+use tokio::net::TcpStream;
 use tokio::sync::{Notify, RwLock};
 
 use task_executor::Executor;
 
+async fn handle_client_output(
+  mut stdio_read: impl Stream<Item = ChildOutput> + Unpin,
+) -> Result<(), io::Error> {
+  let mut stdout = tokio::io::stdout();
+  let mut stderr = tokio::io::stderr();
+  while let Some(output) = stdio_read.next().await {
+    match output {
+      ChildOutput::Stdout(bytes) => stdout.write_all(&bytes).await?,
+      ChildOutput::Stderr(bytes) => stderr.write_all(&bytes).await?,
+      ChildOutput::Exit(_) => {
+        // NB: We ignore exit here and allow the main thread to handle exiting.
+        break;
+      }
+    }
+  }
+  Ok(())
+}
+
+async fn handle_client_input(mut stdin_write: mpsc::Sender<ChildInput>) -> Result<(), io::Error> {
+  use nails::execution::send_to_io;
+  let mut stdin = stream_for(tokio::io::stdin());
+  while let Some(input_bytes) = stdin.next().await {
+    stdin_write
+      .send(ChildInput::Stdin(input_bytes?))
+      .await
+      .map_err(send_to_io)?;
+  }
+  stdin_write
+    .send(ChildInput::StdinEOF)
+    .await
+    .map_err(send_to_io)?;
+  Ok(())
+}
+
+async fn client_execute_helper(
+  port: u16,
+  command: String,
+  args: Vec<String>,
+  env: Vec<(String, String)>,
+) -> Result<i32, String> {
+  use nails::execution::{child_channel, Command};
```

Comment:
I personally like using them, since it keeps import scope small if a given import is only used in a single function.

Reply:
But there is no way to enforce that they are used in a single function, so they'll end up copied pasted elsewhere. And "import scope" only matters if there are imported name collisions... in which case it would be very confusing to use this strategy to override the file-level import (instead you'd want to rename/alias one of the collisions at the file level).